README

  m68k-tester
  A M680x0 emulator tester

  Copyright (C) 2007 Gwenole Beauchesne


License
-------

The m68k-tester program is available under the terms of the GNU
General Public License. See the file "COPYING" that is included in the
distribution for details.


Overview
--------

m68k-tester is aimed at being useful to M680x0 emulator authors to
debug their CPU engine. It's originally designed to cope with results
files generated by Ray Arachelian on his NeXT machine (68040).


Requirements
------------

You need the results files:
<http://gwenole.beauchesne.info/projects/68ktester/files/m68k-tests.tar.bz2>

You need a precompiled M680x0 emulation engine available in object
form (.o files packed into a libemu68k.a static archive) along with
the necessary headers.

Other specific software:
  - GNU make


Quick build instructions
------------------------

Carefully edit the Makefile for emulator specific defines. They have
to match what your emulator's configure script generated. Otherwise,
some data structures won't be of the same size, thus possibly placing
your emulator data to the wrong location.

Define the following variables at the make command line or in the Makefile:
  EMULATOR         name of the emulator (BasiliskII, Aranym, {,E-}UAE, QEMU)
  EMULATOR_PATH    path to the emulator (actually, emulator headers)
  MEMORY           memory addressing mode to use (real, direct, banks)


Acknowledgements
----------------

Special thanks to:
  - Ray Arachelian for composing and providing the m68k tests{,results}.


History
-------

Please consult the file "NEWS" for the release history.
An auto-generated "ChangeLog" is also available in this distribution.


Author
------

Gwenole Beauchesne,
<gb.public@free.fr>


NEWS
m68k-tester NEWS -- history of human-readable changes.  2007-07-08
Copyright (C) 2007 Gwenole Beauchesne

Version 1.0 (SNAPSHOT) - 08.Jul.2007
* First public release


src/config.h
#if defined __LP64__
#define SIZEOF_LONG 8
#define SIZEOF_VOID_P 8
#else
#define SIZEOF_LONG 4
#define SIZEOF_VOID_P 4
#endif

#if defined __BIG_ENDIAN__
#define WORDS_BIGENDIAN 1
#elif defined __linux__
#include <endian.h>
#if __BYTE_ORDER == __BIG_ENDIAN
#define WORDS_BIGENDIAN 1
#endif
#endif

#define HAVE_FCNTL_H 1
#define HAVE_UNISTD_H 1
#define HAVE_SIGACTION 1
#define HAVE_SIGNAL 1
#define HAVE_GETTIMEOFDAY 1

#if defined __APPLE__ && defined __MACH__
#define HAVE_MACH_VM 1
#else
#define HAVE_MMAP_VM 1
#endif

#ifdef HAVE_MACH_VM
#define HAVE_MACH_EXCEPTIONS 1

#define HAVE_VM_ALLOCATE 1
#define HAVE_VM_DEALLOCATE 1
#define HAVE_VM_PROTECT 1
#define HAVE_MACH_TASK_SELF 1
#define HAVE_MACH_VM_MAP_H 1
#endif

#ifdef HAVE_MMAP_VM
#define HAVE_SIGINFO_T 1

#define HAVE_MMAP 1
#define HAVE_MUNMAP 1
#define HAVE_MPROTECT 1
#define HAVE_SYS_MMAN_H 1
#if defined(__linux__)
#define HAVE_MMAP_ANON 1
#define HAVE_MMAP_ANONYMOUS 1
#endif
#endif


src/sysdeps.h
/*
 *  sysdeps.h - System dependent definitions
 *
 *  m68k-tester (C) 2007 Gwenole Beauchesne
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef SYSDEPS_H
#define SYSDEPS_H

#include "config.h"

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <stdint.h>
typedef int8_t int8;
typedef int16_t int16;
typedef int32_t int32;
typedef int64_t int64;
typedef uint8_t uint8;
typedef uint16_t uint16;
typedef uint32_t uint32;
typedef uint64_t uint64;
typedef uintptr_t uintptr;

#endif /* SYSDEPS_H */


src/m68k-tester-qemu.cpp
/*
 *  m68k-tester-qemu.cpp - M68K emulator tester, glue for QEMU-based cores
 *
 *  m68k-tester (C) 2007 Gwenole Beauchesne
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <malloc.h>

#include "sysdeps.h"
#include "vm_alloc.h"
#include "m68k-tester.h"

#define DEBUG 0
#include "debug.h"


/* <config-host.h> glue */
#if defined __LP64__
#define HOST_LONG_BITS 64
#endif
#if defined __linux__
#define HAVE_BYTESWAP_H 1
#endif

/* <config.h> glue */
#define TARGET_M68K 1
#define TARGET_WORDS_BIGENDIAN 1
#if 1
#define CONFIG_USER_ONLY 1
#define CONFIG_LINUX_USER 1
#endif
#define CONFIG_SOFTFLOAT 1

/* <softfloat.h> glue */
#if (defined(__i386__) || defined(__x86_64__)) && !defined(_BSD)
#define FLOATX80
#endif
typedef uint32 float32;
typedef uint64 float64;
typedef struct float_status {
    signed char float_detect_tininess;
    signed char float_rounding_mode;
    signed char float_exception_flags;
#ifdef FLOATX80
    signed char floatx80_rounding_precision;
#endif
} float_status;

#if defined(__x86_64__)
#define TARGET_PHYS_ADDR_BITS 64
#elif defined(__i386__)
#define TARGET_PHYS_ADDR_BITS 32
#endif

#if EMU_QEMU
extern "C" {
#define NEED_CPU_H
#include "cpu.h"
extern void tb_flush(CPUM68KState*);
extern void cpu_dump_state(CPUM68KState *env, FILE *f,
						   int (*cpu_fprintf)(FILE *f, const char *fmt, ...),
						   int flags);
}
#endif

static int m68k_memory_init(void)
{
	if (vm_init() < 0)
		return -1;
	if (vm_acquire_fixed((void *)(uintptr)M68K_CODE_BASE, M68K_CODE_SIZE) < 0)
		return -1;
	return 0;
}

static void m68k_memory_exit(void)
{
	vm_release((void *)(uintptr)M68K_CODE_BASE, M68K_CODE_SIZE);
}

static CPUM68KState *m68k_cpu_init(void)
{
	char cpu_str[] = "m68000";
	cpu_str[4] = CPUType + '0';
	CPUM68KState *cpu = cpu_m68k_init(cpu_str);
	if (cpu == NULL)
		return NULL;

	return cpu;
}

static void m68k_cpu_exit(CPUM68KState *cpu)
{
	cpu_m68k_close(cpu);
}

static void m68k_execute(CPUM68KState *cpu)
{
	for (;;) {
		int trapnr = cpu_m68k_exec(cpu);
		switch (trapnr) {
		case EXCP_EXEC_RETURN:
			// special opcode, exit from m68k execution loop
			return;
		default:
			fprintf(stderr, "qemu: unhandled CPU exception 0x%x - aborting\n",
					trapnr);
			cpu_dump_state(cpu, stderr, fprintf, 0);
			abort();
		}
	}
}


#define M68K_STATE ((CPUM68KState *)opaque)

m68k_cpu::m68k_cpu()
{
	m68k_memory_init();
	cpu_exec_init_all(0);
	opaque = m68k_cpu_init();
	assert(opaque != NULL);
}

m68k_cpu::~m68k_cpu()
{
	m68k_cpu_exit(M68K_STATE);
	m68k_memory_exit();
}

uint32 m68k_cpu::get_pc() const
{
	return M68K_STATE->pc;
}

void m68k_cpu::set_pc(uint32 pc)
{
	M68K_STATE->pc = pc;
}

uint32 m68k_cpu::get_ccr() const
{
	return M68K_STATE->sr & M68K_CCR_BITS;
}

void m68k_cpu::set_ccr(uint32 ccr)
{
	M68K_STATE->sr = (M68K_STATE->sr & ~M68K_CCR_BITS) | (ccr & M68K_CCR_BITS);
}

uint32 m68k_cpu::get_dreg(int r) const
{
	return M68K_STATE->dregs[r];
}

void m68k_cpu::set_dreg(int r, uint32 v)
{
	M68K_STATE->dregs[r] = v;
}

uint32 m68k_cpu::get_areg(int r) const
{
	return M68K_STATE->aregs[r];
}

void m68k_cpu::set_areg(int r, uint32 v)
{
	M68K_STATE->aregs[r] = v;
}

void m68k_cpu::reset(void)
{
}

void m68k_cpu::reset_jit(void)
{
	tb_flush(M68K_STATE);
}

void m68k_cpu::execute(uint32 pc)
{
	D(bug("* execute code at %08x\n", pc));
	set_pc(pc);
	m68k_execute(M68K_STATE);
}

extern "C" {

int singlestep;
unsigned long last_brk;

void pstrcpy(char *buf, int buf_size, const char *str)
{
    int c;
    char *q = buf;

    if (buf_size <= 0)
        return;

    for(;;) {
        c = *str++;
        if (c == 0 || q >= buf + buf_size - 1)
            break;
        *q++ = c;
    }
    *q = '\0';
}

void qemu_free(void *ptr)
{
    free(ptr);
}

void *qemu_malloc(size_t size)
{
    return malloc(size);
}

void *qemu_mallocz(size_t size)
{
    void *ptr;
    ptr = qemu_malloc(size);
    if (!ptr)
        return NULL;
    memset(ptr, 0, size);
    return ptr;
}

void mmap_lock(void )
{
    return;
}

void mmap_unlock(void)
{
    return;
}

void cpu_list_lock(void)
{
}

void *qemu_vmalloc(size_t size)
{
    return memalign(4096, size);
}

void cpu_list_unlock(void)
{
}

typedef int (*gdb_reg_cb)(CPUState *env, uint8_t *buf, int reg);

void gdb_register_coprocessor(CPUState * env,
                             gdb_reg_cb get_reg, gdb_reg_cb set_reg,
                             int num_regs, const char *xml, int g_pos)
{
}
}


src/m68k-tester-dummy.cpp
/*
 *  m68k-tester-dummy.cpp - M68K emulator tester, glue template
 *
 *  m68k-tester (C) 2007 Gwenole Beauchesne
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "sysdeps.h"
#include "vm_alloc.h"
#include "m68k-tester.h"

#define DEBUG 0
#include "debug.h"


m68k_cpu::m68k_cpu()
{
}

m68k_cpu::~m68k_cpu()
{
}

uint32 m68k_cpu::get_pc() const
{
	return 0;
}

void m68k_cpu::set_pc(uint32 pc)
{
}

uint32 m68k_cpu::get_ccr() const
{
	return 0;
}

void m68k_cpu::set_ccr(uint32 ccr)
{
}

uint32 m68k_cpu::get_dreg(int r) const
{
	return 0;
}

void m68k_cpu::set_dreg(int r, uint32 v)
{
}

uint32 m68k_cpu::get_areg(int r) const
{
	return 0;
}

void m68k_cpu::set_areg(int r, uint32 v)
{
}

void m68k_cpu::reset(void)
{
}

void m68k_cpu::reset_jit(void)
{
}

void m68k_cpu::execute(uint32 pc)
{
	D(bug("* execute code at %08x\n", pc));
}

uint32 vm_get_byte(uint32 addr)
{
	return 0;
}

void vm_put_byte(uint32 addr, uint8 v)
{
}

uint32 vm_get_word(uint32 addr)
{
	return 0;
}

void vm_put_word(uint32 addr, uint16 v)
{
}

uint32 vm_get_long(uint32 addr)
{
	return 0;
}

void vm_put_long(uint32 addr, uint32 v)
{
}


src/vm_alloc.cpp
/*
 *  vm_alloc.cpp - Wrapper to various virtual memory allocation schemes
 *                 (supports mmap, vm_allocate or fallbacks to malloc)
 *
 *  m68k-tester (C) 2007 Gwenole Beauchesne
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif

#ifdef HAVE_WIN32_VM
#define WIN32_LEAN_AND_MEAN /* avoid including junk */
#include <windows.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include "vm_alloc.h"

#ifdef HAVE_MACH_VM
#ifndef HAVE_MACH_TASK_SELF
#ifdef HAVE_TASK_SELF
#define mach_task_self task_self
#else
#error "No task_self(), you lose."
#endif
#endif
#endif

/* We want MAP_32BIT, if available, for SheepShaver and BasiliskII
   because the emulated target is 32-bit and this helps to allocate
   memory so that branches could be resolved more easily (32-bit
   displacement to code in .text), on AMD64 for example.  */
#ifndef MAP_32BIT
#define MAP_32BIT 0
#endif
#ifndef MAP_ANON
#define MAP_ANON 0
#endif
#ifndef MAP_ANONYMOUS
#define MAP_ANONYMOUS 0
#endif

#define MAP_EXTRA_FLAGS (MAP_32BIT)

#ifdef HAVE_MMAP_VM
#if (defined(__linux__) && defined(__i386__)) || HAVE_LINKER_SCRIPT
/* Force a reasonnable address below 0x80000000 on x86 so that we
   don't get addresses above when the program is run on AMD64.
   NOTE: this is empirically determined on Linux/x86.  */
#define MAP_BASE	0x10000000
#else
#define MAP_BASE	0x00000000
#endif
static char * next_address = (char *)MAP_BASE;
#ifdef HAVE_MMAP_ANON
#define map_flags	(MAP_ANON | MAP_EXTRA_FLAGS)
#define zero_fd		-1
#else
#ifdef HAVE_MMAP_ANONYMOUS
#define map_flags	(MAP_ANONYMOUS | MAP_EXTRA_FLAGS)
#define zero_fd		-1
#else
#define map_flags	(MAP_EXTRA_FLAGS)
static int zero_fd	= -1;
#endif
#endif
#endif

/* Translate generic VM map flags to host values.  */

#ifdef HAVE_MMAP_VM
static int translate_map_flags(int vm_flags)
{
	int flags = 0;
	if (vm_flags & VM_MAP_SHARED)
		flags |= MAP_SHARED;
	if (vm_flags & VM_MAP_PRIVATE)
		flags |= MAP_PRIVATE;
	if (vm_flags & VM_MAP_FIXED)
		flags |= MAP_FIXED;
	if (vm_flags & VM_MAP_32BIT)
		flags |= MAP_32BIT;
	return flags;
}
#endif

/* Align ADDR and SIZE to 64K boundaries.  */

#ifdef HAVE_WIN32_VM
static inline LPVOID align_addr_segment(LPVOID addr)
{
	return (LPVOID)(((DWORD)addr) & -65536);
}

static inline DWORD align_size_segment(LPVOID addr, DWORD size)
{
	return size + ((DWORD)addr - (DWORD)align_addr_segment(addr));
}
#endif

/* Translate generic VM prot flags to host values.  */

#ifdef HAVE_WIN32_VM
static int translate_prot_flags(int prot_flags)
{
	int prot = PAGE_READWRITE;
	if (prot_flags == (VM_PAGE_EXECUTE | VM_PAGE_READ | VM_PAGE_WRITE))
		prot = PAGE_EXECUTE_READWRITE;
	else if (prot_flags == (VM_PAGE_EXECUTE | VM_PAGE_READ))
		prot = PAGE_EXECUTE_READ;
	else if (prot_flags == (VM_PAGE_READ | VM_PAGE_WRITE))
		prot = PAGE_READWRITE;
	else if (prot_flags == VM_PAGE_READ)
		prot = PAGE_READONLY;
	else if (prot_flags == 0)
		prot = PAGE_NOACCESS;
	return prot;
}
#endif

/* Initialize the VM system. Returns 0 if successful, -1 for errors.  */

int vm_init(void)
{
#ifdef HAVE_MMAP_VM
#ifndef zero_fd
	zero_fd = open("/dev/zero", O_RDWR);
	if (zero_fd < 0)
		return -1;
#endif
#endif
	return 0;
}

/* Deallocate all internal data used to wrap virtual memory allocators.  */

void vm_exit(void)
{
#ifdef HAVE_MMAP_VM
#ifndef zero_fd
	if (zero_fd != -1) {
		close(zero_fd);
		zero_fd = -1;
	}
#endif
#endif
}

/* Allocate zero-filled memory of SIZE bytes. The mapping is private
   and default protection bits are read / write. The return value
   is the actual mapping address chosen or VM_MAP_FAILED for errors.  */

void * vm_acquire(size_t size, int options)
{
	void * addr;

	// VM_MAP_FIXED are to be used with vm_acquire_fixed() only
	if (options & VM_MAP_FIXED)
		return VM_MAP_FAILED;

#ifdef HAVE_MACH_VM
	// vm_allocate() returns a zero-filled memory region
	if (vm_allocate(mach_task_self(), (vm_address_t *)&addr, size, TRUE) != KERN_SUCCESS)
		return VM_MAP_FAILED;
#else
#ifdef HAVE_MMAP_VM
	int fd = zero_fd;
	int the_map_flags = translate_map_flags(options) | map_flags;

	if ((addr = mmap((caddr_t)next_address, size, VM_PAGE_DEFAULT, the_map_flags, fd, 0)) == (void *)MAP_FAILED)
		return VM_MAP_FAILED;
	
	// Sanity checks for 64-bit platforms
	if (sizeof(void *) == 8 && (options & VM_MAP_32BIT) && !((char *)addr <= (char *)0xffffffff))
		return VM_MAP_FAILED;

	next_address = (char *)addr + size;
#else
#ifdef HAVE_WIN32_VM
	if ((addr = VirtualAlloc(NULL, size, MEM_RESERVE | MEM_COMMIT, PAGE_EXECUTE_READWRITE)) == NULL)
		return VM_MAP_FAILED;
#else
	if ((addr = calloc(size, 1)) == 0)
		return VM_MAP_FAILED;
	
	// Omit changes for protections because they are not supported in this mode
	return addr;
#endif
#endif
#endif

	// Explicitely protect the newly mapped region here because on some systems,
	// say MacOS X, mmap() doesn't honour the requested protection flags.
	if (vm_protect(addr, size, VM_PAGE_DEFAULT) != 0)
		return VM_MAP_FAILED;
	
	return addr;
}

/* Allocate zero-filled memory at exactly ADDR (which must be page-aligned).
   Retuns 0 if successful, -1 on errors.  */

int vm_acquire_fixed(void * addr, size_t size, int options)
{
	// Fixed mappings are required to be private
	if (options & VM_MAP_SHARED)
		return -1;

#ifdef HAVE_MACH_VM
	// vm_allocate() returns a zero-filled memory region
	if (vm_allocate(mach_task_self(), (vm_address_t *)&addr, size, 0) != KERN_SUCCESS)
		return -1;
#else
#ifdef HAVE_MMAP_VM
	int fd = zero_fd;
	int the_map_flags = translate_map_flags(options) | map_flags | MAP_FIXED;

	if (mmap((caddr_t)addr, size, VM_PAGE_DEFAULT, the_map_flags, fd, 0) == (void *)MAP_FAILED)
		return -1;
#else
#ifdef HAVE_WIN32_VM
	// Windows cannot allocate Low Memory
	if (addr == NULL)
		return -1;

	// Allocate a possibly offset region to align on 64K boundaries
	LPVOID req_addr = align_addr_segment(addr);
	DWORD  req_size = align_size_segment(addr, size);
	LPVOID ret_addr = VirtualAlloc(req_addr, req_size, MEM_COMMIT | MEM_RESERVE, PAGE_EXECUTE_READWRITE);
	if (ret_addr != req_addr)
		return -1;
#else
	// Unsupported
	return -1;
#endif
#endif
#endif
	
	// Explicitely protect the newly mapped region here because on some systems,
	// say MacOS X, mmap() doesn't honour the requested protection flags.
	if (vm_protect(addr, size, VM_PAGE_DEFAULT) != 0)
		return -1;
	
	return 0;
}

/* Deallocate any mapping for the region starting at ADDR and extending
   LEN bytes. Returns 0 if successful, -1 on errors.  */

int vm_release(void * addr, size_t size)
{
	// Safety check: don't try to release memory that was not allocated
	if (addr == VM_MAP_FAILED)
		return 0;

#ifdef HAVE_MACH_VM
	if (vm_deallocate(mach_task_self(), (vm_address_t)addr, size) != KERN_SUCCESS)
		return -1;
#else
#ifdef HAVE_MMAP_VM
	if (munmap((caddr_t)addr, size) != 0)
		return -1;
#else
#ifdef HAVE_WIN32_VM
	if (VirtualFree(align_addr_segment(addr), 0, MEM_RELEASE) == 0)
		return -1;
#else
	free(addr);
#endif
#endif
#endif
	
	return 0;
}

/* Change the memory protection of the region starting at ADDR and
   extending LEN bytes to PROT. Returns 0 if successful, -1 for errors.  */

int vm_protect(void * addr, size_t size, int prot)
{
#ifdef HAVE_MACH_VM
	int ret_code = vm_protect(mach_task_self(), (vm_address_t)addr, size, 0, prot);
	return ret_code == KERN_SUCCESS ? 0 : -1;
#else
#ifdef HAVE_MMAP_VM
	int ret_code = mprotect((caddr_t)addr, size, prot);
	return ret_code == 0 ? 0 : -1;
#else
#ifdef HAVE_WIN32_VM
	DWORD old_prot;
	int ret_code = VirtualProtect(addr, size, translate_prot_flags(prot), &old_prot);
	return ret_code != 0 ? 0 : -1;
#else
	// Unsupported
	return -1;
#endif
#endif
#endif
}

/* Returns the size of a page.  */

int vm_get_page_size(void)
{
#ifdef HAVE_WIN32_VM
	static unsigned long page_size = 0;
	if (page_size == 0) {
		SYSTEM_INFO si;
		GetSystemInfo(&si);
		page_size = si.dwAllocationGranularity;
	}
	return page_size;
#else
	return getpagesize();
#endif
}

#ifdef CONFIGURE_TEST_VM_MAP
#include <stdlib.h>
#include <signal.h>

static void fault_handler(int sig)
{
	exit(1);
}

/* Tests covered here:
   - TEST_VM_PROT_* program slices actually succeeds when a crash occurs
   - TEST_VM_MAP_ANON* program slices succeeds when it could be compiled
*/
int main(void)
{
	vm_init();

	signal(SIGSEGV, fault_handler);
#ifdef SIGBUS
	signal(SIGBUS,  fault_handler);
#endif
	
#define page_align(address) ((char *)((unsigned long)(address) & -page_size))
	unsigned long page_size = vm_get_page_size();
	
	const int area_size = 6 * page_size;
	volatile char * area = (volatile char *) vm_acquire(area_size);
	volatile char * fault_address = area + (page_size * 7) / 2;

#if defined(TEST_VM_MMAP_ANON) || defined(TEST_VM_MMAP_ANONYMOUS)
	if (area == VM_MAP_FAILED)
		return 1;

	if (vm_release((char *)area, area_size) < 0)
		return 1;
	
	return 0;
#endif

#if defined(TEST_VM_PROT_NONE_READ) || defined(TEST_VM_PROT_NONE_WRITE)
	if (area == VM_MAP_FAILED)
		return 0;
	
	if (vm_protect(page_align(fault_address), page_size, VM_PAGE_NOACCESS) < 0)
		return 0;
#endif

#if defined(TEST_VM_PROT_RDWR_WRITE)
	if (area == VM_MAP_FAILED)
		return 1;
	
	if (vm_protect(page_align(fault_address), page_size, VM_PAGE_READ) < 0)
		return 1;
	
	if (vm_protect(page_align(fault_address), page_size, VM_PAGE_READ | VM_PAGE_WRITE) < 0)
		return 1;
#endif

#if defined(TEST_VM_PROT_READ_WRITE)
	if (vm_protect(page_align(fault_address), page_size, VM_PAGE_READ) < 0)
		return 0;
#endif

#if defined(TEST_VM_PROT_NONE_READ)
	// this should cause a core dump
	char foo = *fault_address;
	return 0;
#endif

#if defined(TEST_VM_PROT_NONE_WRITE) || defined(TEST_VM_PROT_READ_WRITE)
	// this should cause a core dump
	*fault_address = 'z';
	return 0;
#endif

#if defined(TEST_VM_PROT_RDWR_WRITE)
	// this should not cause a core dump
	*fault_address = 'z';
	return 0;
#endif
}
#endif
